loader/src/ui/nodes/TextArea.cpp

```diff
@@ -0,0 +1,229 @@
+}
+
+CCTextAlignment SimpleTextArea::getAlignment() {
+    return this->m_alignment;
+}
+
+void SimpleTextArea::setText(const std::string& text) {
+    this->m_text = text;
+
+    this->updateContents();
+}
+
+std::string SimpleTextArea::getText() {
+    return this->m_text;
+}
+
+void SimpleTextArea::setMaxLines(const size_t maxLines) {
+    this->m_maxLines = maxLines;
+
+    this->updateContents();
+}
+
+size_t SimpleTextArea::getMaxLines() {
+    return this->m_maxLines;
+}
+
+void SimpleTextArea::setWidth(const float width) {
+        this->m_artificialWidth = true;
+
+        this->setContentSize({ width, this->getContentSize().height });
+        this->m_container->setContentSize({ width, this->m_container->getContentSize().height });
+    }
+
+float SimpleTextArea::getWidth() {
+    return this->m_container->getContentSize().width;
+}
+
+void SimpleTextArea::setScale(const float scale) {
+    this->m_scale = scale;
+
+    this->updateContents();
+}
+
+float SimpleTextArea::getScale() {
+    return this->m_scale;
+}
+
+void SimpleTextArea::setLinePadding(const float padding) {
+    this->m_linePadding = padding;
+
+    this->updateContents();
+}
+
+float SimpleTextArea::getLinePadding() {
+    return this->m_linePadding;
+}
+
+std::vector<CCLabelBMFont*> SimpleTextArea::getLines() {
+    return this->m_lines;
+}
+
+float SimpleTextArea::getHeight() {
+    return this->m_container->getContentSize().height;
+}
+
+float SimpleTextArea::getLineHeight() {
+    return this->m_lineHeight;
+}
+
+CCLabelBMFont* SimpleTextArea::createLabel(const std::string& text, const float top) {
+    CCLabelBMFont* label = CCLabelBMFont::create(text.c_str(), this->m_font.c_str());
+
+    label->setScale(this->m_scale);
+    label->setAnchorPoint({ 0, 0 });
+    label->setPosition({ 0, top });
+
+    return label;
+}
+
+CCLabelBMFont* SimpleTextArea::moveOverflow(CCLabelBMFont* line, const char c, const float top) {
+    const std::string text = line->getString();
+    const char back = text.back();
+    const bool lastIsSpace = back == ' ';
+    CCLabelBMFont* newLine = this->createLabel(std::string(!lastIsSpace, back).append(std::string(c != ' ', c)), top);
+
+    if (!lastIsSpace) {
+        if (text[text.size() - 2] == ' ') {
+            line->setString(text.substr(0, text.size() - 1).c_str());
+        } else {
+            line->setString((text.substr(0, text.size() - 1) + '-').c_str());
+        }
+    }
+
+    this->m_lines.push_back(newLine);
+
+    return newLine;
+}
+
+float SimpleTextArea::calculateOffset(CCLabelBMFont* label) {
+    return this->m_linePadding + label->getContentSize().height * this->m_scale;
+}
+
+void SimpleTextArea::updateLines() {
+    float top = 0;
+    CCLabelBMFont* line = this->createLabel("", top);
+    this->m_lines = { line };
+
+    for (const char c : this->m_text) {
+        if (this->m_maxLines && this->m_lines.size() > this->m_maxLines) {
+            std::cout << this->m_lines.size() << std::endl;
```

Comment:
xd

Reply:
whoops, debug thingy I forgot to remove, let me fix rq